File: C11/백준/Silver/6588. 골드바흐의 추측/골드바흐의 추측.c
#include <stdio.h>
#include <stdbool.h>

#define MAX 1000000

int main(void){
	
	int N;
	
	bool isPrime[MAX + 1];
	isPrime[0] = false;
	isPrime[1] = false;
	
	for(int i = 2; i < MAX + 1; i++){
		isPrime[i] = true;
	}
	
	for(int i = 2; i * i < MAX + 1; i++){
		if(isPrime[i]){
			for(int j = i * i; j < MAX + 1; j += i){
				isPrime[j] = false;
			}	
		}
	}
	
	while(scanf("%d", &N) == 1){
		if(N == 0){
			break;
		}
		
		bool found = false;
		
		for(int a = 3; a <= N / 2; a += 2){
			int b = N - a;
			if(isPrime[a] && isPrime[b]){
				printf("%d = %d + %d\n", N, a, b);
				found = true;
				break;
			}
		}
		if(!found){
			printf("Goldbach's conjecture is wrong.");
		}
	}
	
}

File: Java/백준/Silver/6588. 골드바흐의 추측/골드바흐의 추측.java
import java.io.*;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) {

        try {
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

            boolean[] isprime = new boolean[1000001];
            Arrays.fill(isprime, true);
            isprime[0] = isprime[1] = false;

            //에라토스테네스의 체를 이용한 소수 판별
            for (int i = 2; i * i < 1000001; i++) {
                if (isprime[i]) {
                    for (int j = i * i; j < 1000001; j += i) {
                        isprime[j] = false;
                    }
                }
            }

            String line;

            while ((line = br.readLine()) != null){
                int N = Integer.parseInt(line.trim());
                if (N == 0) break; //0이 입력되면 종료된다.

                boolean found = false;

                for (int a = 3; a <= N; a += 2){
                    int b = N - a;
                    if (isprime[a] && isprime[b]) {
                        bw.write(N + " = " + a + " + " + b);
                        bw.newLine();
                        found = true;
                        break;
                    }
                }

                if (!found){
                    bw.write("Goldbach's conjecture is wrong.");
                    bw.newLine();
                }
            }
            bw.flush();
            bw.close();
            br.close();
        } catch (IOException e) { //입출력 예외 처리
            e.printStackTrace();
        } catch (NumberFormatException e) { //숫자 변환시 예외 처리
            e.printStackTrace();
        }
    }
}